### PR TITLE
feat: Allow decoding of non-Polars arrow dictionaries in Arrow and Parquet

### LINKED ISSUE #17945: Polars cannot read native pyarrow parquet tables that use dictionary encodings (e.g., categorical types)
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import pyarrow as pa
import pyarrow.parquet as pq
import polars as pl

schema = pa.schema([
    ('whatever', pa.dictionary(pa.int64(), pa.utf8()))
])

arr_1 = pa.array(["Flamingo", "Parrot", "Dog"]).dictionary_encode()
arr_2 = pa.array(["Horse", "Brittle stars", "Centipede", "Dog"]).dictionary_encode()

num_repeats = 1000000

large_arr_1 = pa.concat_arrays([arr_1] * num_repeats)
large_arr_2 = pa.concat_arrays([arr_2] * num_repeats)

data1 = pa.Table.from_arrays([large_arr_1], schema=schema)
data2 = pa.Table.from_arrays([large_arr_2], schema=schema)

print("data1")
print('_' * 80)
print(data1)
print('_' * 80)
print("data2")
print('_' * 80)
print(data2)
print('_' * 80)

total = pa.concat_tables((data1, data2))

pq.write_table(total, 'sure')

total_back = pq.read_table('sure')

print("total_back")
print('_' * 80)
print(total_back)
print('_' * 80)

print(pl.show_versions())

pl_total = pl.read_parquet('sure')

print("pl_total", pl_total)
```

This returns 
```bash
ComputeError                              Traceback (most recent call last)
[<ipython-input-2-306fa500aaed>](https://localhost:8080/#) in <cell line: 42>()
     40 print(pl.show_versions())
     41 
---> 42 pl_total = pl.read_parquet('sure')
     43 
     44 print("pl_total", pl_total)

3 frames
[/usr/local/lib/python3.10/dist-packages/polars/_utils/deprecation.py](https://localhost:8080/#) in wrapper(*args, **kwargs)
     89                 old_name, new_name, kwargs, function.__qualname__, version
     90             )
---> 91             return function(*args, **kwargs)
     92 
     93         wrapper.__signature__ = inspect.signature(function)  # type: ignore[attr-defined]

[/usr/local/lib/python3.10/dist-packages/polars/_utils/deprecation.py](https://localhost:8080/#) in wrapper(*args, **kwargs)
     89                 old_name, new_name, kwargs, function.__qualname__, version
     90             )
---> 91             return function(*args, **kwargs)
     92 
     93         wrapper.__signature__ = inspect.signature(function)  # type: ignore[attr-defined]

[/usr/local/lib/python3.10/dist-packages/polars/io/parquet/functions.py](https://localhost:8080/#) in read_parquet(source, columns, n_rows, row_index_name, row_index_offset, parallel, use_statistics, hive_partitioning, glob, hive_schema, try_parse_hive_dates, rechunk, low_memory, storage_options, retries, use_pyarrow, pyarrow_options, memory_map)
    206             lf = lf.select(columns)
    207 
--> 208     return lf.collect()
    209 
    210 

[/usr/local/lib/python3.10/dist-packages/polars/lazyframe/frame.py](https://localhost:8080/#) in collect(self, type_coercion, predicate_pushdown, projection_pushdown, simplify_expression, slice_pushdown, comm_subplan_elim, comm_subexpr_elim, cluster_with_columns, no_optimization, streaming, engine, background, _eager, **_kwargs)
   2024         # Only for testing purposes
   2025         callback = _kwargs.get("post_opt_callback", callback)
-> 2026         return wrap_df(ldf.collect(callback))
   2027 
   2028     @overload

ComputeError: parquet: File out of specification: OOM: Decoding ByteArray "Plain"-encoded optional  parquet pages not yet implemented
```

### Log output

_No response_

### Issue description

If data is encoded in a pyarrow dictionary type, it should be read in by polars into a categorical column, not throw an error. Thanks to @EthanSteinberg for the MWE

### Expected behavior

It should read in the file, not throw an error. Note that this works if you pass `use_pyarrow=True` during the read, but it should be supported either way by the engine. 

E.g., this works, and should be the expected behavior for the other syntax too:

```python
import pyarrow as pa
import pyarrow.parquet as pq
import polars as pl

schema = pa.schema([
    ('whatever', pa.dictionary(pa.int64(), pa.utf8()))
])

arr_1 = pa.array(["Flamingo", "Parrot", "Dog"]).dictionary_encode()
arr_2 = pa.array(["Horse", "Brittle stars", "Centipede", "Dog"]).dictionary_encode()

num_repeats = 1000000

large_arr_1 = pa.concat_arrays([arr_1] * num_repeats)
large_arr_2 = pa.concat_arrays([arr_2] * num_repeats)

data1 = pa.Table.from_arrays([large_arr_1], schema=schema)
data2 = pa.Table.from_arrays([large_arr_2], schema=schema)

print("data1")
print('_' * 80)
print(data1)
print('_' * 80)
print("data2")
print('_' * 80)
print(data2)
print('_' * 80)

total = pa.concat_tables((data1, data2))

pq.write_table(total, 'sure')

total_back = pq.read_table('sure')

print("total_back")
print('_' * 80)
print(total_back)
print('_' * 80)

print(pl.show_versions())

pl_total = pl.read_parquet('sure', use_pyarrow=True)

print("pl_total", pl_total)
```

This prints
```bash
data1
________________________________________________________________________________
pyarrow.Table
whatever: dictionary<values=string, indices=int64, ordered=0>
----
whatever: [  -- dictionary:
["Flamingo","Parrot","Dog"]  -- indices:
[0,1,2,0,1,...,1,2,0,1,2]]
________________________________________________________________________________
data2
________________________________________________________________________________
pyarrow.Table
whatever: dictionary<values=string, indices=int64, ordered=0>
----
whatever: [  -- dictionary:
["Horse","Brittle stars","Centipede","Dog"]  -- indices:
[0,1,2,3,0,...,3,0,1,2,3]]
________________________________________________________________________________
total_back
________________________________________________________________________________
pyarrow.Table
whatever: dictionary<values=string, indices=int32, ordered=0>
----
whatever: [  -- dictionary:
["Flamingo","Parrot","Dog"]  -- indices:
[0,1,2,0,1,...,0,1,2,0,1],  -- dictionary:
["Flamingo","Parrot","Dog"]  -- indices:
[2,0,1,2,0,...,2,0,1,2,0],...,  -- dictionary:
["Horse","Brittle stars","Centipede","Dog"]  -- indices:
[0,1,2,3,0,...,3,0,1,2,3],  -- dictionary:
["Horse","Brittle stars","Centipede","Dog"]  -- indices:
[0,1,2,3,0,...,3,0,1,2,3]]
________________________________________________________________________________
--------Version info---------
Polars:               1.3.0
Index type:           UInt32
Platform:             Linux-6.1.85+-x86_64-with-glibc2.35
Python:               3.10.12 (main, Mar 22 2024, 16:50:05) [GCC 11.4.0]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          2.2.1
connectorx:           <not installed>
deltalake:            <not installed>
fastexcel:            <not installed>
fsspec:               2024.6.1
gevent:               <not installed>
great_tables:         <not installed>
hvplot:               <not installed>
matplotlib:           3.7.1
nest_asyncio:         1.6.0
numpy:                1.26.4
openpyxl:             3.1.5
pandas:               2.1.4
pyarrow:              14.0.2
pydantic:             2.8.2
pyiceberg:            <not installed>
sqlalchemy:           2.0.31
torch:                2.3.1+cu121
xlsx2csv:             <not installed>
xlsxwriter:           <not installed>
None
pl_total shape: (7_000_000, 1)
┌───────────────┐
│ whatever      │
│ ---           │
│ cat           │
╞═══════════════╡
│ Flamingo      │
│ Parrot        │
│ Dog           │
│ Flamingo      │
│ Parrot        │
│ …             │
│ Dog           │
│ Horse         │
│ Brittle stars │
│ Centipede     │
│ Dog           │
└───────────────┘
```

### Installed versions

<details>

```
--------Version info---------
Polars:               1.3.0
Index type:           UInt32
Platform:             Linux-6.1.85+-x86_64-with-glibc2.35
Python:               3.10.12 (main, Mar 22 2024, 16:50:05) [GCC 11.4.0]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          2.2.1
connectorx:           <not installed>
deltalake:            <not installed>
fastexcel:            <not installed>
fsspec:               2024.6.1
gevent:               <not installed>
great_tables:         <not installed>
hvplot:               <not installed>
matplotlib:           3.7.1
nest_asyncio:         1.6.0
numpy:                1.26.4
openpyxl:             3.1.5
pandas:               2.1.4
pyarrow:              14.0.2
pydantic:             2.8.2
pyiceberg:            <not installed>
sqlalchemy:           2.0.31
torch:                2.3.1+cu121
xlsx2csv:             <not installed>
xlsxwriter:           <not installed>```

</details>

### LINKED ISSUE #20242: Panic When Aggregating Parquet LazyFrame: validity mask length must match the number of values
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl 
import datetime
import zoneinfo 
print(pl.__version__)
start_date = datetime.datetime(2024, 10, 2, 0, 0, tzinfo=zoneinfo.ZoneInfo(key='UTC'))
df = pl.scan_parquet("polars_minimal_repro.parquet")
df.filter(pl.col("datetime_col").gt(start_date)).group_by("string_col").agg(
    pl.n_unique("enum_col"),
).collect()
```

In polars 1.12.0, this is fine: 
```
>>> import polars as pl 
>>> import datetime
>>> import zoneinfo 
>>> print(pl.__version__)
1.12.0
>>> start_date = datetime.datetime(2024, 10, 2, 0, 0, tzinfo=zoneinfo.ZoneInfo(key='UTC'))
>>> df = pl.scan_parquet("polars_minimal_repro.parquet")
>>> df.filter(pl.col("datetime_col").gt(start_date)).group_by("string_col").agg(
...     pl.n_unique("enum_col"),
... ).collect()
shape: (7_897_465, 2)
┌──────────────────────┬──────────┐
│ string_col           ┆ enum_col │
│ ---                  ┆ ---      │
│ str                  ┆ u32      │
╞══════════════════════╪══════════╡
│ 8604122513921788285  ┆ 1        │
│ 6028805287010285681  ┆ 1        │
│ 5325329696969523616  ┆ 1        │
│ 17510005466429574147 ┆ 1        │
│ 12290517284928471532 ┆ 1        │
│ …                    ┆ …        │
│ 2236822817610814946  ┆ 1        │
│ 7255832370614837454  ┆ 1        │
│ 18220362948434179914 ┆ 1        │
│ 14065194591260067140 ┆ 1        │
│ 14544684152382932294 ┆ 1        │
└──────────────────────┴──────────┘
```

On the latest polars, it fails:
```
Python 3.12.2 (main, Feb  6 2024, 20:19:44) [Clang 15.0.0 (clang-1500.1.0.2.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import polars as pl 
>>> import datetime
>>> import zoneinfo 
>>> print(pl.__version__)
1.17.1
>>> start_date = datetime.datetime(2024, 10, 2, 0, 0, tzinfo=zoneinfo.ZoneInfo(key='UTC'))
>>> df = pl.scan_parquet("polars_minimal_repro.parquet")
>>> df.filter(pl.col("datetime_col").gt(start_date)).group_by("string_col").agg(
...     pl.n_unique("enum_col"),
... ).collect()
thread 'polars-5' panicked at /Users/runner/work/polars/polars/crates/polars-arrow/src/array/primitive/mod.rs:447:48:
called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("validity mask length must match the number of values"))
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'polars-10' panicked at /Users/runner/work/polars/polars/crates/polars-arrow/src/array/primitive/mod.rs:447:48:
called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("validity mask length must match the number of values"))
thread 'polars-9' panicked at /Users/runner/work/polars/polars/crates/polars-arrow/src/array/primitive/mod.rs:447:48:
called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("validity mask length must match the number of values"))
thread 'polars-4' panicked at /Users/runner/work/polars/polars/crates/polars-arrow/src/array/primitive/mod.rs:447:48:
called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("validity mask length must match the number of values"))
thread 'polars-9' panicked at /Users/runner/work/polars/polars/crates/polars-arrow/src/array/primitive/mod.rs:447:48:
called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("validity mask length must match the number of values"))
thread 'polars-0' panicked at /Users/runner/work/polars/polars/crates/polars-arrow/src/array/primitive/mod.rs:447:48:
called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("validity mask length must match the number of values"))
thread 'polars-9' panicked at /Users/runner/work/polars/polars/crates/polars-arrow/src/array/primitive/mod.rs:447:48:
called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("validity mask length must match the number of values"))
thread 'polars-5' panicked at /Users/runner/work/polars/polars/crates/polars-arrow/src/array/primitive/mod.rs:447:48:
called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("validity mask length must match the number of values"))
thread 'polars-4' panicked at /Users/runner/work/polars/polars/crates/polars-arrow/src/array/primitive/mod.rs:447:48:
called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("validity mask length must match the number of values"))
thread 'polars-6' panicked at /Users/runner/work/polars/polars/crates/polars-arrow/src/array/primitive/mod.rs:447:48:
called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("validity mask length must match the number of values"))
thread 'polars-10' panicked at /Users/runner/work/polars/polars/crates/polars-arrow/src/array/primitive/mod.rs:447:48:
called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("validity mask length must match the number of values"))
thread 'polars-8' panicked at /Users/runner/work/polars/polars/crates/polars-arrow/src/array/primitive/mod.rs:447:48:
called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("validity mask length must match the number of values"))
thread 'polars-7' panicked at /Users/runner/work/polars/polars/crates/polars-arrow/src/array/primitive/mod.rs:447:48:
called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("validity mask length must match the number of values"))
thread 'polars-5' panicked at /Users/runner/work/polars/polars/crates/polars-arrow/src/array/primitive/mod.rs:447:48:
called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("validity mask length must match the number of values"))
thread 'polars-2' panicked at /Users/runner/work/polars/polars/crates/polars-arrow/src/array/primitive/mod.rs:447:48:
called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("validity mask length must match the number of values"))
thread 'polars-5' panicked at /Users/runner/work/polars/polars/crates/polars-arrow/src/array/primitive/mod.rs:447:48:
called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("validity mask length must match the number of values"))
thread 'polars-5' panicked at /Users/runner/work/polars/polars/crates/polars-arrow/src/array/primitive/mod.rs:447:48:
called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("validity mask length must match the number of values"))
thread 'polars-11' panicked at /Users/runner/work/polars/polars/crates/polars-arrow/src/array/primitive/mod.rs:447:48:
called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("validity mask length must match the number of values"))
thread 'polars-3' panicked at /Users/runner/work/polars/polars/crates/polars-arrow/src/array/primitive/mod.rs:447:48:
called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("validity mask length must match the number of values"))
thread 'polars-6' panicked at /Users/runner/work/polars/polars/crates/polars-arrow/src/array/primitive/mod.rs:447:48:
called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("validity mask length must match the number of values"))
Traceback (most recent call last):
  File "<stdin>", line 3, in <module>
  File "/Users/aryik/Library/Caches/pypoetry/virtualenvs/LFRtXiTt-py3.12/lib/python3.12/site-packages/polars/lazyframe/frame.py", line 2031, in collect
    return wrap_df(ldf.collect(callback))
                   ^^^^^^^^^^^^^^^^^^^^^
pyo3_runtime.PanicException: called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("validity mask length must match the number of values"))
```

I'm unable to attach `polars_minimal_repro.parquet` here because it's 560 MB. Please let me know what a good way is to share the file, I'm happy to do so and made an effort to create an obfuscated version of the offending parquet file. The size of the file seems to be closely related to the issue - a truncated version of the file does not fail. 

### Log output

```shell
Python 3.12.2 (main, Feb  6 2024, 20:19:44) [Clang 15.0.0 (clang-1500.1.0.2.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
Python 3.12.2 (main, Feb  6 2024, 20:19:44) [Clang 15.0.0 (clang-1500.1.0.2.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
keys/aggregates are not partitionable: running default HASH AGGREGATION
keys/aggregates are not partitionable: running default HASH AGGREGATION
parquet scan with parallel = Prefiltered
parquet scan with parallel = Prefiltered
parquet live columns = 1, dead columns = 2
parquet live columns = 1, dead columns = 2
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet file must be read, statistics not sufficient for predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet file must be read, statistics not sufficient for predicate.
parquet file must be read, statistics not sufficient for predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet row group must be read, statistics not sufficient for predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet file must be read, statistics not sufficient for predicate.
parquet file must be read, statistics not sufficient for predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet row group must be read, statistics not sufficient for predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet file must be read, statistics not sufficient for predicate.
parquet file must be read, statistics not sufficient for predicate.
parquet row group must be read, statistics not sufficient for predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet file must be read, statistics not sufficient for predicate.
parquet row group must be read, statistics not sufficient for predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet file must be read, statistics not sufficient for predicate.
parquet file must be read, statistics not sufficient for predicate.
parquet file must be read, statistics not sufficient for predicate.
parquet file must be read, statistics not sufficient for predicate.
parquet row group must be read, statistics not sufficient for predicate.
parquet row group must be read, statistics not sufficient for predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet file must be read, statistics not sufficient for predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet file must be read, statistics not sufficient for predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet row group must be read, statistics not sufficient for predicate.
parquet file must be read, statistics not sufficient for predicate.
parquet row group must be read, statistics not sufficient for predicate.
parquet row group must be read, statistics not sufficient for predicate.
parquet row group must be read, statistics not sufficient for predicate.
parquet file must be read, statistics not sufficient for predicate.
parquet row group must be read, statistics not sufficient for predicate.
parquet row group must be read, statistics not sufficient for predicate.
parquet file must be read, statistics not sufficient for predicate.
parquet file must be read, statistics not sufficient for predicate.
parquet row group must be read, statistics not sufficient for predicate.
parquet row group must be read, statistics not sufficient for predicate.
parquet row group must be read, statistics not sufficient for predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet row group must be read, statistics not sufficient for predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet file must be read, statistics not sufficient for predicate.
parquet row group must be read, statistics not sufficient for predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet file must be read, statistics not sufficient for predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet file must be read, statistics not sufficient for predicate.
parquet row group must be read, statistics not sufficient for predicate.
parquet row group must be read, statistics not sufficient for predicate.
parquet row group must be read, statistics not sufficient for predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet row group must be read, statistics not sufficient for predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet file must be read, statistics not sufficient for predicate.
parquet row group must be read, statistics not sufficient for predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet file can be skipped, the statistics were sufficient to apply the predicate.
parquet file must be read, statistics not sufficient for predicate.
parquet row group can be skipped, the statistics were sufficient to apply the predicate.
parquet file must be read, statistics not sufficient for predicate.
parquet row group must be read, statistics not sufficient for predicate.
parquet row group must be read, statistics not sufficient for predicate.
thread 'polars-5' panicked at /Users/runner/work/polars/polars/crates/polars-arrow/src/array/primitive/mod.rs:447:48:
called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("validity mask length must match the number of values"))
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'polars-5' panicked at /Users/runner/work/polars/polars/crates/polars-arrow/src/array/primitive/mod.rs:447:48:
called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("validity mask length must match the number of values"))
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


TRUNCATED DUE TO GITHUB LIMITATIONS. Full file is attached to the description. 
```

### Issue description

We first ran into this issue on Polars 1.13, and are currently stuck on Polars 1.12 because of it. We have a production system that heavily utilizes polars. As a part of the system, we have to write out a very large parquet file occasionally - this file contains ~50,000,000 rows of data. Upon an upgrade to polars 1.13, we started seeing a few issues which I originally reported on Discord. 

This issue is a `PanicException`:
```
Traceback (most recent call last):
  File "<stdin>", line 3, in <module>
  File "/Users/aryik/Library/Caches/pypoetry/virtualenvs/decision-engine-LFRtXiTt-py3.12/lib/python3.12/site-packages/polars/lazyframe/frame.py", line 2031, in collect
  File "<stdin>", line 3, in <module>
  File "/Users/aryik/Library/Caches/pypoetry/virtualenvs/decision-engine-LFRtXiTt-py3.12/lib/python3.12/site-packages/polars/lazyframe/frame.py", line 2031, in collect
    return wrap_df(ldf.collect(callback))
    return wrap_df(ldf.collect(callback))
                   ^^^^^^^^^^^^^^^^^^^^^
                   ^^^^^^^^^^^^^^^^^^^^^
pyo3_runtime.PanicException: called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("validity mask length must match the number of values"))
```

I've been trying to isolate the issue, but struggled to figure out exactly what's going on. Here are some notes:

- It only seems to happen with `scan_parquet` .
- It seems to have something to do with aggregations on enum columns.
- It seems to have something to do with filtering on datetimes with timezones.
- It seems to pass when I turn optimization off.
- It only seems to happen on large parquet files - when we run the exact same code on a smaller dataset, it passes. 

Prior to polars 1.17, I would get a panic exception on `pl.scan("other_minimal_repro.parquet").head().collect()`, but one of @coastalwhite's fixes seems to have resolved that issue. 

In case it's helpful, I've attached the log output from the successful run on polars 1.12: 
[stderr_polars_1_12_success.txt](https://github.com/user-attachments/files/18072676/stderr_polars_1_12_success.txt)

I've also attached the full output from 1.17.1: 
[stderr_polars_1_17_failure.txt](https://github.com/user-attachments/files/18072667/stderr_polars_1_17_failure.txt)

### Expected behavior

I expect to be able to filter and aggregate without getting a `PanicException`. 

### Installed versions

<details>
Note: I also observed this on an x64 linux EC2 machine. 

```
--------Version info---------
Polars:              1.17.1
Index type:          UInt32
Platform:            macOS-15.1-arm64-arm-64bit
Python:              3.12.2 (main, Feb  6 2024, 20:19:44) [Clang 15.0.0 (clang-1500.1.0.2.5)]
LTS CPU:             False

----Optional dependencies----
adbc_driver_manager  <not installed>
altair               <not installed>
boto3                1.35.76
cloudpickle          3.1.0
connectorx           <not installed>
deltalake            <not installed>
fastexcel            <not installed>
fsspec               <not installed>
gevent               <not installed>
google.auth          2.36.0
great_tables         <not installed>
matplotlib           3.9.3
nest_asyncio         1.6.0
numpy                2.0.2
openpyxl             <not installed>
pandas               2.2.3
pyarrow              18.1.0
pydantic             2.10.2
pyiceberg            <not installed>
sqlalchemy           <not installed>
torch                <not installed>
xlsx2csv             <not installed>
xlsxwriter           <not installed>
```

</details>

### LINKED ISSUE #20270: Creating a Polars dataframe from an arrow dictionary with null values produces an invalid categorical categories
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl
import pyarrow as pa

t = pa.table(
    {
        "col1": pa.DictionaryArray.from_arrays([0, 0, None, 1, 2], ["A", None, "B"]),
    },
    schema=pa.schema({"col1": pa.dictionary(pa.uint32(), pa.string()) }),
)
assert pl.from_arrow(t).select(pl.col.col1.cat.get_categories().null_count()).to_series().item() == 0
```

### Log output

_No response_

### Issue description

Categorical categories contain nulls

### Expected behavior

Categorical categories should not contain nulls

### Installed versions

<details>

```
Replace this line with the output of pl.show_versions(). Leave the backticks in place.
```

</details>

### LINKED ISSUE #20271: `from_arrow` does not respect global StringCache
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
with pl.StringCache():
    s = pl.Series('a', ["A", "B", "C"], pl.Categorical)
    df = pl.from_arrow(pa.table({ 'b': pa.DictionaryArray.from_arrays([0, 1], ["D", "E"]) }))
    print(s)
    print(s.to_physical())
    print()
    print(df.to_series())
    print(df.to_series().to_physical())
```

### Log output

_No response_

### Issue description

Physicals are local

### Expected behavior

`from_arrow` respects the StringCache

### Installed versions

<details>

```
Replace this line with the output of pl.show_versions(). Leave the backticks in place.
```

</details>

### LINKED ISSUE #20288: Categorical lexical ordering is not kept when roundtripped through arrow or parquet
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl
import io

f = io.BytesIO()
df = pl.Series('a', ["A", "B"], pl.Categorical(ordering='lexical')).to_frame()
df.write_parquet(f)
f.seek(0)
pl.scan_parquet(f).collect_schema()['a'] # Output: Categorical(ordering='physical')
```

### Log output

_No response_

### Issue description

Ordering is saved in arrow schema

### Expected behavior

Ordering to be kept

### Installed versions

<details>

```
Replace this line with the output of pl.show_versions(). Leave the backticks in place.
```

</details>

### PR DESCRIPTION
**Pull request by coastalwhite**
This removes all instances of the DictionaryEncoder except one which is used for Polars enums and categoricals. This essentially makes it so that the dictionary arrow type is regarded as any other arrow type.

Fixes #20242.
Fixes #17945.
Fixes #20270.
Fixes #20288.
Fixes #20271.

This should massively speed up the decoding of enums and categoricals, although that is very much not the goal. This PR unifies the decoder kernels and removes a lot of the useless monomorphizations.

### CODE REVIEW

**Comment by coastalwhite** [comment]
Blocked on #20250.

**Comment by coastalwhite** [comment]
After some consideration, `from_arrow` now also converts dictionaries to flat arrays if they cannot put into a Categorical.

**Comment by coastalwhite** [comment]
Okay, soooooo. This turned into a way larger change than I originally had in mind, but I still think this can be a relatively cohesive patch-set. In the end, I really wanted to things:

1. Make sure that the `ArrowDataType::Dictionary` conversion is consistent. This is why both Parquet and Arrow code had to be hit.
2. Allow to keep all Polars DataType information when converting to and from Parquet and Arrow.

The rules for converting from an `ArrowDataType::Dictionary` are now as follows.
1. Check the Metadata for `_PL_ENUM_VALUES` and `_PL_CATEGORICAL`? If either is available, load the enum values or the categorical ordering, respectively. For Parquet, this also informs the reader the column can read using by just looking at the dictionary encoded indexes. This saves a lot of memory consumption and saves a big group-by. Categoricals properly follow the `StringCache` if it is enabled.
2. If there is no metadata and the value datatype is a `String`, cast to a Categorical with `physical` ordering. This now also properly utilizes the StringCache i.f.f. it is enabled. We also now properly propagate nulls from the dictionary values onto the dictionary keys.
3. If the value datatype is a non-string datatype, convert the dictionary to plain values.

Similarly, when we convert a Categorical or Enum to arrow, we now set `_PL_ENUM_VALUES` with the enum values (was already done) and `_PL_CATEGORICAL` with the ordering (was not yet done). To properly roundtrip through arrow this metadata had to be added to the `RecordBatch`. I did this with a `ArrowSchemaRef`. This way, `to_arrow()` and `from_arrow` can now properly roundtrip categoricals and enums. I think the same from IPC and Avro, although I did not check this yet.

File: crates/polars-compute/src/propagate_dictionary.rs [hunk 1]
```diff
@@ -0,0 +1,87 @@
+use arrow::array::{Array, BinaryViewArray, PrimitiveArray, Utf8ViewArray};
+use arrow::bitmap::Bitmap;
+use arrow::datatypes::ArrowDataType::UInt32;
+
+/// Propagate the nulls from the dictionary values into the keys and remove those nulls from the
+/// values.
+pub fn propagate_dictionary_value_nulls(
```
Comment (ritchie46) [inline crates/polars-compute/src/propagate_dictionary.rs:7]: Ah, good one.

File: crates/polars-core/src/frame/mod.rs [hunk 2]
```diff
@@ -3255,6 +3271,7 @@ impl DataFrame {
 
 pub struct RecordBatchIter<'a> {
     columns: &'a Vec<Column>,
+    schema: ArrowSchemaRef,
```
Comment (ritchie46) [inline crates/polars-core/src/frame/mod.rs:3274]: Why do we want the schema on `RecordBatch`? Do we want O(1) access somewhere?

**Review by ritchie46 — APPROVED** [review]
Will already

File: crates/polars-core/src/frame/mod.rs (hunk 2 above)
Comment (coastalwhite) [reply crates/polars-core/src/frame/mod.rs:3274]: We need it to be able to properly round trip types. We need the ArrowFields, the ArrowDataType isn't enough.

Comment (ritchie46) [reply crates/polars-core/src/frame/mod.rs:3274]: Ah right. Convinces me more that we should shift to `Fields` on `Series`/`Columns`.

**Pull request merged**